File: 207.Course Schedule.cpp
/*
 * @lc app=leetcode id=207 lang=cpp
 * @lcpr version=20004
 *
 * [207] Course Schedule
 *
 * https://leetcode.com/problems/course-schedule/description/
 *
 * algorithms
 * Medium (48.11%)
 * Likes:    16700
 * Dislikes: 770
 * Total Accepted:    1.9M
 * Total Submissions: 3.9M
 * Testcase Example:  '2\n[[1,0]]'
 *
 * There are a total of numCourses courses you have to take, labeled from 0 to
 * numCourses - 1. You are given an array prerequisites where prerequisites[i]
 * = [ai, bi] indicates that you must take course bi first if you want to take
 * course ai.
 * 
 * 
 * For example, the pair [0, 1], indicates that to take course 0 you have to
 * first take course 1.
 * 
 * 
 * Return true if you can finish all courses. Otherwise, return false.
 * 
 * 
 * Example 1:
 * 
 * Input: numCourses = 2, prerequisites = [[1,0]]
 * Output: true
 * Explanation: There are a total of 2 courses to take. 
 * To take course 1 you should have finished course 0. So it is possible.
 * 
 * 
 * Example 2:
 * 
 * Input: numCourses = 2, prerequisites = [[1,0],[0,1]]
 * Output: false
 * Explanation: There are a total of 2 courses to take. 
 * To take course 1 you should have finished course 0, and to take course 0 you
 * should also have finished course 1. So it is impossible.
 * 
 * 
 * 
 * Constraints:
 * 
 * 
 * 1 <= numCourses <= 2000
 * 0 <= prerequisites.length <= 5000
 * prerequisites[i].length == 2
 * 0 <= ai, bi < numCourses
 * All the pairs prerequisites[i] are unique.
 * 
 * 
 */


// @lcpr-template-start
using namespace std;
#include <algorithm>
#include <array>
#include <bitset>
#include <climits>
#include <deque>
#include <functional>
#include <iostream>
#include <list>
#include <queue>
#include <stack>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
// @lcpr-template-end
// @lc code=start
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> table(numCourses);
        for(auto& pre : prerequisites){
            table[pre[1]].push_back(pre[0]);
      
        }

        vector<int> memo(numCourses,0);
        vector<int> s(numCourses,0);

        for(int i = 0; i < numCourses;i++){
            if(dfs(table,memo,i,s) == false) return false;
        }
        return true;
    }

    bool dfs(vector<vector<int>>& table, vector<int>& memo, int start,vector<int>& visited){
        if(memo[start] == 1)return true;
        if(memo[start] == -1) return false;
        cout<<start<<endl;
        for(auto i : table[start]){
            if(visited[i] == 1) {
                memo[start] = -1;
                return false;
            }
            visited[i] = 1;
            if(dfs(table,memo,i,visited) == false) {
                memo[start] = -1;
                return false;
            }
            visited[i] = 0;
        }

        memo[start] = 1;
        return true;
    }
};
// @lc code=end



/*
// @lcpr case=start
// 2\n[[1,0]]\n
// @lcpr case=end

// @lcpr case=start
// 2\n[[1,0],[0,1]]\n
// @lcpr case=end

 */

File: 207.Course Schedule.py
#
# @lc app=leetcode id=207 lang=python3
# @lcpr version=20004
#
# [207] Course Schedule
#
# https://leetcode.com/problems/course-schedule/description/
#
# algorithms
# Medium (48.23%)
# Likes:    16760
# Dislikes: 772
# Total Accepted:    1.9M
# Total Submissions: 3.9M
# Testcase Example:  '2\n[[1,0]]'
#
# There are a total of numCourses courses you have to take, labeled from 0 to
# numCourses - 1. You are given an array prerequisites where prerequisites[i] =
# [ai, bi] indicates that you must take course bi first if you want to take
# course ai.
# 
# 
# For example, the pair [0, 1], indicates that to take course 0 you have to
# first take course 1.
# 
# 
# Return true if you can finish all courses. Otherwise, return false.
# 
# 
# Example 1:
# 
# Input: numCourses = 2, prerequisites = [[1,0]]
# Output: true
# Explanation: There are a total of 2 courses to take. 
# To take course 1 you should have finished course 0. So it is possible.
# 
# 
# Example 2:
# 
# Input: numCourses = 2, prerequisites = [[1,0],[0,1]]
# Output: false
# Explanation: There are a total of 2 courses to take. 
# To take course 1 you should have finished course 0, and to take course 0 you
# should also have finished course 1. So it is impossible.
# 
# 
# 
# Constraints:
# 
# 
# 1 <= numCourses <= 2000
# 0 <= prerequisites.length <= 5000
# prerequisites[i].length == 2
# 0 <= ai, bi < numCourses
# All the pairs prerequisites[i] are unique.
# 
# 
#


# @lcpr-template-start

# @lcpr-template-end
# @lc code=start
class UF:
    def __init__(self,n):
        self.parent = list(range(n))
    
    def find(self,x):
        if x != self.parent[x]:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    def union(self,x,y):
        a = self.find(x)
        b = self.find(y)
        if a != b:
            self.parent[b] = a
            
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        map = defaultdict(list)
        for a, b in prerequisites:
            map[b].append(a)
        visited = [0 for _ in range(numCourses)]
            
        def dfs(map,i):
            if visited[i] == 1:
                return False
            if visited[i] == 2:
                return True
            visited[i] = 1
            for a in map[i]:
                if not dfs(map,a):
                    return False
                
            visited[i] = 2
            return True
        
        for y in range(numCourses):
            if not dfs(map,y):
                return False
        return True


            
# @lc code=end



#
# @lcpr case=start
# 2\n[[1,0]]\n
# @lcpr case=end

# @lcpr case=start
# 2\n[[1,0],[0,1]]\n
# @lcpr case=end

#

